File: go/collatz-conjecture/collatz_conjecture.go
package collatzconjecture

import "errors"

func CollatzConjecture(n int) (int, error) {
    if n <= 0 {
        return 0, errors.New("Argument must be a positive integer")
    }
    numberOfSteps := 0
	for n != 1 {
        numberOfSteps++
        if n % 2 == 0{
            n = n / 2
        } else {
        	n = 3 * n + 1
        }
    }
	return numberOfSteps, nil
}

File: cpp/collatz-conjecture/collatz_conjecture.cpp
#include "collatz_conjecture.h"
#include <stdexcept>

namespace collatz_conjecture {
    int steps(int num) {
        if (num < 1) {
            throw std::domain_error("input must be positive");
        } else if (num == 1) {
            return 0;
        } else {
            return 1 + (num % 2 == 0 ? steps(num / 2) : steps(3 * num + 1));
        }
    }
}  // namespace collatz_conjecture
